File: zad3_1.py
import os

class Logger():
    def __init__(self):
        self.history=[]
    def history_add(self, textToLog):
        self.history.append(textToLog)
    def log(self, textToLog):
        self.history_add(textToLog)
    def __del__(self):
        pass

class FileLogger(Logger):
    def __init__(self, filePath):
        self.history=[]
        self.logFile=open(filePath, "a+")
    def log(self, textToLog):
        self.history_add(textToLog)
        self.logFile.write(textToLog + "\n")
    def __del__(self):
        self.logFile.close()

class StdoutLogger(Logger):
    def log(self, textToLog):
        self.history_add(textToLog)
        print(textToLog)


def saveLogs(loggers, text):
    for log in loggers:
        log.log(text)

def stopLogging(loggers):
    for log in loggers:
        del log 


dir=os.path.join(os.path.dirname(__file__), "logs")

loggers=[StdoutLogger(), 
        FileLogger(os.path.join(dir, "log1.txt")),
        FileLogger(os.path.join(dir, "log2.txt"))]

texts=["123", "456", "789",
    "abc", "qwe"]

for line in texts:
    saveLogs(loggers, line)

stopLogging(loggers)

File: zad3_1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
using namespace std;

class Logger{
    public:
        virtual void log(string textToLog)=0;
        virtual void del(){};

        vector<string> history;
        
    protected:
        virtual void history_add(string textToLog){
            history.push_back(textToLog);
        }
};

class FileLogger : public Logger{
    public:
        FileLogger(string filePath){
            logFile.open(filePath);
        }
        void log(string textToLog){
            history_add(textToLog);
            logFile << textToLog << "\n";
        }
        void del(){
            logFile.close();
        }
    private:
        ofstream logFile;
};

class StdoutLogger : public Logger{
    public:
        void log(string textToLog){
            history_add(textToLog);
            cout << textToLog << "\n";
        }
};

void saveLogs(vector<Logger*> &logs, string text){
    for(Logger* l : logs)
        l->log(text);
}

void stopLogging(vector<Logger*> &logs){
    for(Logger* l : logs)
        l->del();
}

int main(void){
    vector<Logger*> logs={new FileLogger("/home/marcin/Documents/Coding/POB_2019/logs/log1.txt"),
                        new FileLogger("/home/marcin/Documents/Coding/POB_2019/logs/log2.txt"),
                        new StdoutLogger()};
    vector<string> texts={"cpp1", "cpp2", "ilovecpp", "linuxoverwindows"};
    for(string text : texts)
        saveLogs(logs, text);
    stopLogging(logs);

}



